Add validation for offset pagination (Schedule A, B, E)

lbeaufort opened the pull request:
## Summary (required)

Resolves #4343

- **Add pagination validation:** For Schedules A, B, and E, if only `last_index` and not any of the sort options or `sort_hide_null=True` is chosen, throw a 422 error. Add test coverage.
- **Fix some Schedule A and Schedule B tests**: Make sure they pass both index values. For some tests, passing only `last_index` was hiding null values, which doesn't really make sense. Passing just `last_index` won't be valid behavior anymore, so for those tests, I explicitly added the `sort_hide_null=True` filter.

**Pagination basics** 
(from https://api.open.fec.gov/developers/#/receipts/get_schedules_schedule_a_):
> The `/schedules​/schedule_a​/` endpoint is not paginated by page number. This endpoint uses keyset pagination to improve query performance and these indices are required to properly page through this large dataset. To request the next page, you should append the values found in the last_indexes object from pagination to the URL of your last request as additional parameters. For example, when sorting by `contribution_receipt_date`, you might receive a page of results with the following pagination information:
```
pagination: {
    pages: 2152643,
    per_page: 20,
    count: 43052850,
    last_indexes: {
        last_index: "230880619",
        last_contribution_receipt_date: "2014-01-01"
    }
}
```
>To fetch the next page of sorted results, append `last_index=230880619` and `last_contribution_receipt_date=2014-01-01` to the URL. We strongly advise paging through these results using sort indices. The default sort is ascending by contribution_receipt_date. If you do not page using sort indices, some transactions may be unintentionally filtered out.

## How to test the changes locally

- Paginate through Schedule A, B, and E. This is tough to give exact links for because the data is changing all the time - the easiest way is to start with these queries and then copy/paste `last_index` and `last_contribution_receipt_date` or `sort_null_only=True` from the `pagination. last_indexes` object into new queries.
- Schedule A ascending: http://localhost:5000/v1/schedules/schedule_a/?two_year_transaction_period=2020&sort=contribution_receipt_date
- Schedule A descending: http://localhost:5000/v1/schedules/schedule_a/?two_year_transaction_period=2020&sort=-contribution_receipt_date
- Schedule B ascending: http://localhost:5000/v1/schedules/schedule_b/?sort=disbursement_date
- Schedule B ascending: http://localhost:5000/v1/schedules/schedule_b/?sort=-disbursement_date
- Schedule E ascending: http://localhost:5000/v1/schedules/schedule_e/?sort=expenditure_date
- Schedule E descending: http://localhost:5000/v1/schedules/schedule_e/?sort=-expenditure_date
- Try sorting/paginating by other `sort_options` as defined in the resource class
- Missing `last_contribution_receipt date` should throw 422 error: http://localhost:5000/v1/schedules/schedule_a/?sort_hide_null=false&sort_nulls_last=false&data_type=processed&min_date=05%2F06%2F2019&max_date=05%2F12%2F2020&sort=-contribution_receipt_date&per_page=100&contributor_name=Jeffrey+Allen&contributor_state=CA&last_index=4030220201231984113
- Test out some Schedule A, B, E queries in Swagger

## Impacted areas of the application
List general components of the application that this PR will affect:

-  Pagination for Schedules A, B, E. Communication costs

lbeaufort commented:
This PR is already pretty big, so I put in this follow-up issue for a usability issue I identified: https://github.com/fecgov/openFEC/issues/4351

fec-jli commented:
Thanks for your great work. I test locally, pytest passed. api Swagger works well locally. 
Here are some more test urls I used:
##1)schedule_a validation:
1)pass last_index and miss the last_sort_column, get error message:
http://127.0.0.1:5000/v1/schedules/schedule_a/?per_page=20&sort_hide_null=false&sort=contribution_receipt_date&sort_null_only=false&contributor_name=Jeffrey%20Allen&last_index=3061920110004692965

2)correct one(with both last_index and one last_sort_column):
http://127.0.0.1:5000/v1/schedules/schedule_a/?per_page=20&sort_hide_null=false&sort=contribution_receipt_date&sort_null_only=false&contributor_name=Jeffrey%20Allen&last_index=3061920110004692965&last_contribution_receipt_date=1995-08-31T00:00:00


##2)schedule_b validation
1)pass last_index and miss the last_sort_column, get error message:
http://127.0.0.1:5000/v1/schedules/schedule_b/?per_page=20&sort_hide_null=false&recipient_name=Paul&sort_null_only=false&sort=-disbursement_date&last_index=4050720201742086704

2)correct one(with both last_index and one last_sort_column):
http://127.0.0.1:5000/v1/schedules/schedule_b/?per_page=20&sort_hide_null=false&recipient_name=Paul&sort_null_only=false&sort=-disbursement_date&last_index=4050720201742086704&last_disbursement_date=2020-03-31T00:00:00

3)miss last_index and pass one last_sort_column, get result (may not pagination correctly)
http://127.0.0.1:5000/v1/schedules/schedule_b/?per_page=20&sort_hide_null=false&recipient_name=Paul&sort_null_only=false&sort=-disbursement_date&last_disbursement_date=2020-03-31T00:00:00

##3)schedule_e valication
1)pass last_index and miss the last_sort_column, get error message:
http://127.0.0.1:5000/v1/schedules/schedule_e/?per_page=20&sort_hide_null=false&payee_name=Paul&sort_null_only=false&sort=-expenditure_date&sort_nulls_last=false&last_index=4110520181608662396

2)correct one(with both last_index and last_sort_column):
http://127.0.0.1:5000/v1/schedules/schedule_e/?per_page=20&sort_hide_null=false&payee_name=Paul&sort_null_only=false&sort=-expenditure_date&sort_nulls_last=false&last_expenditure_date=2018-11-03T00:00:00&last_index=4110520181608662396

3)get result (no pagination)
http://127.0.0.1:5000/v1/schedules/schedule_e/?per_page=20&sort_hide_null=false&payee_name=Paul&sort_null_only=false&sort=-expenditure_date&sort_nulls_last=true

4)sort_null_only=true, pass last_index only. correct result
http://127.0.0.1:5000/v1/schedules/schedule_e/?per_page=20&sort_hide_null=false&payee_name=Paul&sort_null_only=true&sort=-expenditure_date&sort_nulls_last=false&last_index=4110520181608662396

================
question about error message:
1)last_index should be `last_indexes`?
2)The message only show the first sort_column name with last_xxxxx. the rest of them miss last_
 should use for loop ? or write last_column name into sort_options in each resource file.

lbeaufort commented:
@fec-jli thanks for your review! Great catch, I addressed this with c3a7ccd6293711c5929d10700e2e43ba2bb45c9a. 

>1)last_index should be last_indexes?
Agreed, fixed thanks!
>2)The message only show the first sort_column name with last_xxxxx. the rest of them miss last_
should use for loop ? or write last_column name into sort_options in each resource file.
Great catch, I fixed the message. Example: 
http://127.0.0.1:5000/v1/schedules/schedule_e/?per_page=20&sort_hide_null=false&payee_name=Paul&sort=-expenditure_date&sort_nulls_last=false&last_index=4110520181608662396
```
"When paginating through results, both values from the previous page's `last_indexes` object are needed. For more information, see https://api.open.fec.gov/developers/. Please add one of the following filters to your query: `sort_null_only`=True, `last_expenditure_date`, `last_expenditure_amount`, `last_office_total_ytd`, `last_support_oppose_indicator`"
```
